metapool/prep.py

```diff
@@ -938,3 +907,178 @@ def demux_pre_prep(pre_prep):
+    files = glob(f"{fp}/**/filtered_sequences/*.fastq.gz", recursive=True)
+
+    by_project = defaultdict(list)
+
+    run_prefix_mapping = {}
+
+    count = 0
+
+    for fastq_fp in files:
+        # generate the full path to each file for reference.
+        full_path = abspath(fastq_fp)
+
+        # determine the run_prefix and orientation of the file.
+        # We do not want to silently ignore the file if its orientation is not
+        # R1 or R2.
+        run_prefix, orientation = get_run_prefix(basename(fastq_fp))
+        if orientation not in ['R1', 'R2']:
+            raise ValueError(f"{fastq_fp} does not appear to be a forward or "
+                             "reverse read file")
+
+        # process the relative path of each file to ensure each is of the
+        # the expected form. Report anything out of the ordinary so that no
+        # files are silently missed.
+
+        # remove fp from the beginning of the file's path. The remaining path
+        # should be of the form:
+        # <project_name_w_qiita_id>/filtered_sequences/<fastq_file>
+        tmp = fastq_fp.replace(fp, '')
+        # remove any leading and/or trailing '/' characters from the
+        # remaining path.
+        # use sep instead of '/' to be more platform independent.
+        tmp = tmp.strip(sep)
+        tmp = tmp.split(sep)
+
+        # tmp[1] != 'filtered_sequences' doesn't appear to be possible given
+        # the glob statement above, but we'll keep it here for safety.
+        if len(tmp) != 3 or tmp[1] != 'filtered_sequences':
+            raise ValueError(f"{fastq_fp} appears to be stored in an "
+                             "unexpected location")
+
+        # where tmp[0] is the project_name, keep a list of all associated with
+        # that project.
+        by_project[tmp[0]].append(full_path)
+        run_prefix_mapping[full_path] = run_prefix
+        count += 1
+
+    if count == 0:
+        raise ValueError(f"No fastq.gz files were found in {fp}. Please "
+                         "ensure that the path is correct and that the "
+                         "directory contains fastq.gz files.")
+
+    # convert to standard dict before returning.
+    return dict(by_project), run_prefix_mapping
+
+
+def get_run_prefix(file_name):
+    # Modified from mg-scripts/util.py:determine_orientation().
+    orientations = ['R1', 'R2', 'I1', 'I2']
+
+    results = []
+
+    # assume orientation is always present in the file's name.
+    # assume that it is of one of the four forms above.
+    # assume that it is always the right-most occurance of the four
+    # orientations above.
+    # assume that orientation is encapsulated with either '_' or '.'
+    # e.g.: '_R1_', '.I2.'.
+    # assume users can and will include any or all of the four
+    # orientation as part of their filenames as well. e.g.:
+    # ABC_7_04_1776_R1_SRE_S3_L007_R2_001.trimmed.fastq.gz
+    for o in orientations:
+        variations = [f"_{o}_", f".{o}."]
```

Comment:
should this be a global?

Reply:
Do you mean the variations variable or get_run_prefix? The variations variable seems fine where it is. This function is actually a modified version of the determine_orientation() function I added to mg-scripts a while back. This version is meant to be a drop-in replacement for get_run_prefix that also happens to return the orientation so it's a little specialized.

If determine_orientation(filename) is needed in the future we can push it down into metapool.